### PR TITLE
[draft] fix web

### PR DESCRIPTION
**Pull request by munrocket**
I am fixed web version with Emscripten, it works.
But need changes for native build.

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -366,22 +408,34 @@ void run() {
...
+        scDesc.height = kHeight;
         scDesc.presentMode = wgpu::PresentMode::Fifo;
         swapChain = device.CreateSwapChain(surface, &scDesc);
+
+        {
+            wgpu::TextureDescriptor descriptor{};
+            descriptor.usage = wgpu::TextureUsage::RenderAttachment;
+            descriptor.size = {kWidth, kHeight, 1};
+            descriptor.format = wgpu::TextureFormat::Depth32Float;
+            canvasDepthStencilView = device.CreateTexture(&descriptor).CreateView();
```
Comment (kainino0x) [inline main.cpp:421]: I prefer to keep this example minimal so there's less code to dig through. Do you think it's important to add a depth attachment?

File: main.cpp [hunk 2]
```diff
@@ -366,22 +408,34 @@ void run() {
...
+        }
     }
     emscripten_set_main_loop(frame, 0, false);
-#else
-    while (testsCompleted < kNumTests) {
-        device.Tick();
-    }
-#endif
```
Comment (kainino0x) [inline main.cpp:379]: This was for Dawn. Does Dawn no longer need it?

File: main.cpp [hunk 3]
```diff
@@ -366,22 +408,34 @@ void run() {
...
         device = dev;
         run();
     });
-}
+
+    // The test result will be reported when the main_loop completes.
+    // emscripten_exit_with_live_runtime isn't needed because the WebGPU
+    // callbacks should all automatically keep the runtime alive until
+    // emscripten_set_main_loop, and that should keep it alive until
+    // emscripten_cancel_main_loop.
```
Comment (kainino0x) [inline main.cpp:437]: Nice documentation!

**Review by kainino0x — COMMENTED** [review]
This is nice, thanks! Unfortunately I don't have time now to see about the native build right now. I have no idea if it will even work anymore. Do you think you'd be able to look into it at least for Dawn? You'll need to update the Dawn revision which is hardcoded in `setup_native_build.sh`.

**Comment by kainino0x** [comment]
If the answer is the native build against Dawn needs nontrivial Cmake work then let me know and I'll try to find time sometime. It won't be for at least a few weeks though.

File: main.cpp (hunk 2 above)
Comment (munrocket) [reply main.cpp:379]: Yea, I will try to fix it. Was naive and happy that it works.

File: main.cpp (hunk 3 above)
Comment (munrocket) [reply main.cpp:437]: Not mine, copy/paste from emscripten example.

**Comment by shrekshao** [comment]
I had an addition commit that fix the native build: https://github.com/kainino0x/webgpu-cross-platform-demo/commit/d880f5151beb0b3ea4bbaeccd32265a3f711e2f6
(branch on my fork: https://github.com/shrekshao/webgpu-cross-platform-demo/tree/munrocket/main)

With the additional commit it could build for emscripten and native, tested on linux.
Launching the native `hello` will ends in the following error on linux though
```
Error: eglCreateContext failed with EGL_BAD_ATTRIBUTE
    at CheckEGL (/usr/local/google/home/shrekshao/my/webgpu-cross-platform-demo/third_party/dawn/src/dawn/native/opengl/UtilsEGL.cpp:71)
    at Create (/usr/local/google/home/shrekshao/my/webgpu-cross-platform-demo/third_party/dawn/src/dawn/native/opengl/ContextEGL.cpp:85)
```

**Pull request merged**